lib/talks/operations/create.rb

```diff
@@ -65,7 +66,7 @@ def find_or_create_event(event_form)
         if event
           Success(event)
         else
-          new_event = event_repo.create(**event_form)
+          new_event = event_repo.create(**event_form) # state `unpublished` by default
```

Comment:
I rewrite this logic to this one:

```ruby
# frozen_string_literal: true

module Talks
  module Operations
    class Create < Operation
      include Import[
        oembed: 'oembed',
        talk_repo: 'repositories.talk',
        speaker_repo: 'repositories.speaker',
        event_repo: 'repositories.event',
        talks_speakers_repo: 'repositories.talks_speakers',
      ]

      def call(talk_form)
        oembed = yield generate_oembed(talk_form.delete(:link))

        talk_repo.transaction do
          speaker = yield find_or_create_speaker(talk_form[:speaker])
          event   = yield find_or_create_event(talk_form[:event])
          talk    = yield create_talk(talk_form, oembed, event.id)
          yield create_talk_speaker(talk.id, speaker.id)
        end

        Success(talk)
      end

      private

      def generate_oembed(link)
        Try(OEmbed::Error) { oembed.get(link).html }
      end

      # move to approve operation?
      def create_talk_speaker(talk_id, speaker_id)
        talk_speaker = talks_speakers_repo.create(talk_id: talk_id, speaker_id: speaker_id)
        talk_speaker ? Success(talk_speaker) : Failure(:could_not_create_talk_speaker)
      end

      def create_talk(talk_form, oembed, event_id)
        talk = talk_repo.create(**talk_form, embed_code: oembed, event_id: event_id) # state `unpublished` by default
        talk ? Success(talk) : Failure(:could_not_create_talk)
      end

      def find_or_create_speaker(speaker_form)
        speaker = speaker_repo.find_or_create(speaker_form)
        speaker ? Success(speaker) : Failure('could not create speaker')
      end

      def find_or_create_event(event_form)
        event = event_repo.find_or_create(name: event_form[:name])
        event ? Success(event) : Failure('could not create event')
      end
    end
  end
end

# find or create logic - is logic from  repository

class SpeackerRepository
  def find_or_create(speaker_form)
    speaker = find_by_name(first_name: speaker_form[:first_name], last_name: speaker_form[:last_name])
    speaker || create(**speaker_form) # state `unpublished` by default
  end
end

class EventRepository
  def find_or_create(event_form)
    event = find_by_name(name: event_form[:name])
    event || create(**event_form)
  end
end
```

WDYT?

Reply:
the general idea - move create or find logic to repository class. Repository can be fat if you need to implemet fat logic around creating something. 

That's why I also think that transaction should be in repository (or may be in command object, but it's complicated now)